### ISSUE TITLE
Redesign Observer supervisor + worker model

### ISSUE THREAD

**Issue by einari** [Cratis/Chronicle#880]
Today our model is somewhat convoluted from the `Subscribe` perspective.
All logic resides in this one method that will then decide what state the observer is in and what it should be.
This needs to be cleaned up and we should look at it as a statemachine and all the actual work should be separated out into workers that do specific tasks. 

The state in which an observer is in should then be decided by the workers it has working for it. Some worker types require exclusivity and puts the observer in a state were it can't process events from the stream, such as a CatchUp or Replay. While other workers require exclusivity on a partition, such as Failed Partition workers. 

The workers should be stateful and hold a state on their own that gets persisted for the worker. The supervisor state might need to include workers state, meaning that the provider could potentially aggregate information from multiple places for the worker types that are supported and let that be served as a read only state for the supervisor. This would enable the supervisor to make decisions and we would avoid duplicating state that is in reality owned by a worker and not the supervisor. The supervisor makes decisions on the state, but does not create or update it.

As part of the redesign, we also need to look into the model we have today for long-running processes. We're using a timer that gets us out of the worker triggering method (Start) and get called on a separate task context. However, the timer is just another grain call in the eyes of Orleans, meaning that long running worker jobs is a violation in terms of time it spends. 
It would be interesting to use something like [this](https://github.com/Kritner/Orleans.SyncWork). We would need to be able to call grains from it however.

**Comment by einari** [comment]
Things to remember: 

- [ ] Exponential backoff for retry attempts of a failed partition

**Comment by einari** [comment]
- [ ] Job storage
- [ ] Job step storage
- [ ] Specs for Job system
- [ ] Add event sequence number range for HandleEventsForPartition
- [ ] Specs for HandleEventsForPartition job step
- [ ] CatchUp state
- [ ] CatchUp job
- [ ] Retry failed partition job
- [ ] Fix failed partition deep-link navigation from observers list
- [ ] Workbench view for seeing jobs in an event store

**Comment by einari** [comment]
- [ ] Fast forward observers when there are no events between `NextEventSequenceNumber` and tail of the event sequence that it is of a type it is interested in.
- [ ] Disconnect all observers / reducers / projections when shutting down - update their state.
- [ ] Rehydrate all observers on startup

**Comment by einari** [comment]
- [ ] `NextEventSequenceNumberForEventTypes` property needs to be populated on `ObserverState`

**Comment by einari** [comment]
- [ ] When resuming JobSteps - they crash, we're not getting the grain by its type and we just get a generic grain, which is not of type `IJobStep`
- [ ] Add failed partition attempt limit - "giving up"
- [ ] Error handling in the different stages of a job
- [ ] JobsManager should keep track of running jobs in memory. Add a method like `GetRunningJobsOfType<>() and fix the `Replay` state of an observer to use this.
- [ ] Consolidate observer state based on running jobs. If the `RunningState` is set to `Replay` but there are no jobs actually running, we should not be in `Replay`.
- [ ] Jobs that does not have any steps should go straight to `OnCompleted`
- [ ] Specs for `MongoDBObserverKeysAsyncEnumerator`

**Comment by einari** [comment]
- [ ] Before starting `ReplayObserverPartition`, we should set the number of handled events to all without the partition.
